apps/src/dance/ai/utils/chooseEffects.ts

```diff
@@ -23,7 +33,9 @@ const NUM_RANDOM_BOTTOM_OPTIONS = 20;
  */
 export function chooseEffects(
   selectedEmojis: string[],
-  quality: EffectsQuality
+  quality: EffectsQuality,
+  calculateSummedWeights: CalculateSummedWeights = calculateOutputSummedWeights,
+  numOptions: number | undefined = undefined
```

Comment:
Rather than having testing-only args, you could also stub the util function directly in your test - something like `sinon.stub(utils, "calculateOutputSummedWeights")`. 

I know this pattern is used for testing (and I've used it quite a bit too) but as far as I know it's typically used for stubbing constructor parameters, in a similar way to `@VisibleForTesting` constructor overloads in Java. I haven't really seen it used for functions - you've documented it well here so maybe it's not an issue, but it does feel a little weird to me?

Additionally, might be better to have explicit args for the number of top and bottom options with default values (ex. `(..., numTopOptions = NUM_RANDOM_TOP_OPTIONS, numBottomOptions = NUM_RANDOM_BOTTOM_OPTIONS)`)? That way we don't have separate behavior based on if the `numOptions` param was provided or not.

Reply:
Re: sinon stubbing, that was my instinct too...it looks like there's not really a way to directly stub an imported function like we do here, but there are some hacks suggested in this thread that might make it possible:

https://github.com/sinonjs/sinon/issues/562

Reply:
Also waffled on how to parameterize the number of options piece, was thinking maybe a number of options should be provided alongside the type of results you wanted (eg, `{type: 'bad', numOptions: 5}`. Seemed kind of complicated though -- the approach you're suggesting felt a little weird too b/c you'd have to provide a number of top options as an argument even if you're actually just getting bottom options?

Reply:
Yeah re: numOptions, it all feels a little weird I think because of the magic top/bottom options numbers...though I feel like with providing both numbers, it's true that only one will be used based on effects quality, but the caller doesn't necessarily know how the internals work? They're just providing configuration values that this function can use as it needs...alternatively, you could export the `NUM_RANDOM` constants and check against those in your tests?

Reply:
Fair enough 🙂 removing the function argument does make it clearer as well. The only other thing I'd say is that having a single numOptions param then doesn't test the case of returning a different number of options based on effect quality? But I'm okay keeping the single param for clarity

Reply:
Alrighty, I updated to remove numOptions...I mostly put the param there b/c I didn't want to have to write 20 pieces of test data, but I realized that we don't actually need 20 pieces of test data to use the default params. LMK what y'all think!

Reply:
Looks good! Thanks for iterating on this.